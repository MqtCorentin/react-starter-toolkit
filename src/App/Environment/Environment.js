import React, { useState, useEffect } from 'react';
import { Loader } from '@axa-fr/react-toolkit-all';

const NODE_ENV = process.env.REACT_APP_NODE_ENV || process.env.NODE_ENV;
export const { Consumer: EnvironmentConsumer, Provider: EnvironmentProvider } = React.createContext();

/**
 * HOC pass the environment context props to Component
 * @param {*} Component
 */
export const withEnvironment = Component => props => (
  <EnvironmentConsumer>{store => (!store.environment ? <Loader /> : <Component {...props} {...store} />)}</EnvironmentConsumer>
);

const initState = {
  environment: null,
  error: null,
};

export const getImport = () => import(`../../../public/environment.${NODE_ENV}.json`);

/**
 * Hook to setState environment
 */
export const useEnv = (getImportFn = getImport, initStateCt = initState) => {
  const [envState, setEnvState] = useState(initStateCt);

  useEffect(() => {
    if (envState.environment === null) {
      (async () => {
        try {
          const environment = await getImportFn();
          setEnvState({ environment });
        } catch (error) {
          setEnvState({ error });
        }
      })();
    }
  }, [envState.environment, getImportFn]);

  return { envState, setEnvState };
};

/**
 * Component wrapper to add environment context
 * @param {Component} children
 */
const Environment = ({ children }) => {
  const { envState } = useEnv();
  return <EnvironmentProvider value={envState}>{children}</EnvironmentProvider>;
};

export default Environment;
