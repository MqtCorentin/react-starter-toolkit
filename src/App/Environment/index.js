export { default, withEnvironment, EnvironmentProvider, useEnv } from './Environment';
