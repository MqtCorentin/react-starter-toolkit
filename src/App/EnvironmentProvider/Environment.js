import React, { useState, useEffect } from 'react';

const NODE_ENV = process.env.REACT_APP_NODE_ENV || process.env.NODE_ENV;

export const EnvironmentContext = React.createContext();
const { Provider } = EnvironmentContext;

const initState = {
  environment: null,
  error: null,
};

export const getImport = () => import(`../../../public/environment.${NODE_ENV}.json`);

/**
 * Hook to setState environment
 */
export const useEnv = (getImportFn = getImport, initStateCt = initState) => {
  const [envState, setEnvState] = useState(initStateCt);
  useEffect(() => {
    if (envState.environment === null) {
      getImportFn()
        .then(environment => setEnvState({ environment }))
        .catch(error => setEnvState({ error }));
    }
  }, [envState.environment, getImportFn]);

  return { envState, setEnvState };
};

/**
 * Component wrapper to add environment context
 * @param {Component} children
 */
const EnvironmentProvider = ({ children, useEnvFn = useEnv }) => {
  const { envState } = useEnvFn();
  return <Provider value={envState}>{children}</Provider>;
};

export default EnvironmentProvider;
