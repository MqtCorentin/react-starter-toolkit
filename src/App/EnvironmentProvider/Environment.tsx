import { useState, useEffect, createContext, Dispatch, SetStateAction } from 'react';

export type TEnvironment = {
  apiUrl: string;
  baseUrl: string;
  fetchConfig: object;
  oidc: Record<string, boolean | string>;
};

export type TEnvironmentState = {
  environment?: null | TEnvironment;
  error?: unknown;
};

const initState: TEnvironmentState = {
  environment: null,
  error: null,
};

export const EnvironmentContext = createContext<TEnvironmentState>(initState);
EnvironmentContext.displayName = 'EnvironmentContext';

export const getFileEnv = (nodeEnv = process.env.REACT_APP_NODE_ENV || process.env.NODE_ENV) => `environment.${nodeEnv}.json`;

type TfetchEnv = {
  setEnvState: Dispatch<SetStateAction<TEnvironmentState>>;
  signal: AbortSignal;
  fetchFn?: typeof fetch;
  getFileEnvFn?: typeof getFileEnv;
};

export const fetchEnv = async ({ setEnvState, signal, fetchFn = fetch, getFileEnvFn = getFileEnv }: TfetchEnv) => {
  const fileName = getFileEnvFn();
  const response = await fetchFn(`/${fileName}`, { signal }).catch(error => setEnvState({ error }));
  setEnvState({ environment: await response?.json() });
};

/**
 * Hook to setState environment
 */
export const useEnv = (fetchEnvFn = fetchEnv, initStateCt = initState) => {
  const [envState, setEnvState] = useState(initStateCt);
  useEffect(() => {
    const abortController = new AbortController();
    if (envState.environment === null) {
      fetchEnvFn({ setEnvState, signal: abortController.signal });
    }
    return () => abortController.abort();
  }, [envState.environment, fetchEnvFn]);

  return { envState, setEnvState };
};

/**
 * Component wrapper to add environment context
 * @param {Component} children
 */
const EnvironmentProvider = ({ children, useEnvFn = useEnv }: { children: JSX.Element; useEnvFn?: typeof useEnv }) => {
  const { envState } = useEnvFn();
  return <EnvironmentContext.Provider value={envState}>{children}</EnvironmentContext.Provider>;
};

export default EnvironmentProvider;
