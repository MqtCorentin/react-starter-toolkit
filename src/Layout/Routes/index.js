export { default, WithAuth } from './Routes';
