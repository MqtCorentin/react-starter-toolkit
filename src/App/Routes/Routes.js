import React, { useContext } from 'react';
import { func } from 'prop-types';
import { Routes, Route, Navigate } from 'react-router-dom';
import PageNotFound from 'pages/NotFound';
import PageUnauthorize from 'pages/Unauthorize';
import Members from 'pages/Members';
import SearchMembers from 'pages/SearchMembers';
import SlashDesignSystem from 'pages/SlashDesignSystem';
import Modal from 'pages/Demos/Modal';
import Button from 'pages/Demos/Button';
import Dashboard from 'pages/Dashboard';
import ROUTE_URL from 'Layout/constants';
import { UserContext } from 'App/UserProvider';

export const WithAuth = ({ Component, NavigateCmpt = Navigate, authorized = [''], UserContextObj = UserContext }) => {
  const userContext = useContext(UserContextObj);
  return authorized.includes(userContext?.authRole) || !userContext.isEnabled ? <Component /> : <NavigateCmpt to={ROUTE_URL.UNAUTHORIZE} />;
};

const RoutesCmpt = ({
  SlashDesignSystemCmpt,
  DashboardCmpt,
  MembersCmpt,
  SearchMembersCmpt,
  ModalCmpt,
  ButtonCmpt,
  PageUnauthorizeCmpt,
  WithAuthCmpt,
}) => (
  <Routes>
    <Route exact path={ROUTE_URL.SLASH} element={<WithAuthCmpt Component={SlashDesignSystemCmpt} />} />
    <Route exact path={ROUTE_URL.MEMBERS} element={<WithAuthCmpt Component={MembersCmpt} />} />
    <Route exact path={ROUTE_URL.SEARCHMEMBERS} element={<WithAuthCmpt Component={SearchMembersCmpt} />} />
    <Route exact path={ROUTE_URL.MODAL} element={<WithAuthCmpt Component={ModalCmpt} />} />
    <Route exact path={ROUTE_URL.BUTTON} element={<WithAuthCmpt Component={ButtonCmpt} />} />
    <Route exact path={ROUTE_URL.DASHBOARD} element={<WithAuthCmpt Component={DashboardCmpt} />} />
    <Route exact path={ROUTE_URL.UNAUTHORIZE} element={<PageUnauthorizeCmpt />} />
    <Route path="*" element={<PageNotFound />} />
  </Routes>
);

RoutesCmpt.propTypes = {
  SlashDesignSystemCmpt: func,
  DashboardCmpt: func,
  MembersCmpt: func,
  SearchMembersCmpt: func,
  ModalCmpt: func,
  ButtonCmpt: func,
  PageUnauthorizeCmpt: func,
  WithAuthCmpt: func,
};

RoutesCmpt.defaultProps = {
  SlashDesignSystemCmpt: SlashDesignSystem,
  DashboardCmpt: Dashboard,
  MembersCmpt: Members,
  SearchMembersCmpt: SearchMembers,
  ModalCmpt: Modal,
  ButtonCmpt: Button,
  PageUnauthorizeCmpt: PageUnauthorize,
  WithAuthCmpt: WithAuth,
};

export default RoutesCmpt;
