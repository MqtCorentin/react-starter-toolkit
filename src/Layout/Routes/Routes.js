import React from 'react';
import { Switch, Route } from 'react-router-dom';
import { withOidcSecure } from '@axa-fr/react-oidc-context';
import PageNotFound from 'NotFound';
import PageUnauthorize from 'Unauthorize';
import Members from 'Members';
import SlashDesignSystem from 'SlashDesignSystem';
import Dashboard from 'Dashboard';
import Layout from 'Layout';
import ROUTE_URL from 'Layout/constants';

export const renderLayoutMembers = parentProps => props => Layout(Members, props, parentProps);
export const renderLayoutDashboard = parentProps => props => Layout(Dashboard, props, parentProps);
export const renderLayoutSlashDesignSystem = parentProps => props => Layout(SlashDesignSystem, props, parentProps);

const Routes = props => (
  <Switch>
    <Route exact path={ROUTE_URL.DASHBOARD} render={withOidcSecure(renderLayoutDashboard(props))} />
    <Route exact path={ROUTE_URL.MEMBERS} render={withOidcSecure(renderLayoutMembers(props))} />
    <Route exact path={ROUTE_URL.SLASH} render={withOidcSecure(renderLayoutSlashDesignSystem(props))} />
    <Route exact path={ROUTE_URL.UNAUTHORIZE} component={PageUnauthorize} />
    <Route component={PageNotFound} />
  </Switch>
);

export default Routes;
