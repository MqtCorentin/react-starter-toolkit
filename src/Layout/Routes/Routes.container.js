import React, { useContext } from 'react';
import { UserContext } from 'App/User';
import Routes from './Routes';

const RouteWithUserInfo = () => {
  const userContext = useContext(UserContext);
  return <Routes {...userContext} />;
};

export default RouteWithUserInfo;
