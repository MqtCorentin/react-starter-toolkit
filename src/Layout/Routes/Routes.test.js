import React from 'react';
import { BrowserRouter as Router, MemoryRouter } from 'react-router-dom';
import { render, act } from '@testing-library/react';
import Environment from 'App/Environment';
import MembersPage from 'Members';
import DashboardPage from 'Dashboard';
import FormPage from 'Form';
import { renderLayoutMembers, renderLayoutDashboard, renderLayoutForm } from './Routes';
import Routes from '.';

describe('Route', () => {
  it('Should render Routes', async () => {
    await act(async () => {
      const { asFragment } = render(
        <Environment>
          <Router>
            <Routes />
          </Router>
        </Environment>,
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it('NotFound Page', () => {
    const { asFragment } = render(
      <MemoryRouter initialEntries={['/random']}>
        <Routes />
      </MemoryRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('Should return <MembersPage /> with props when call renderLayoutMembers', () => {
    const customProps = {
      myProp: 'property members',
    };
    const result = renderLayoutMembers({ ...customProps });
    expect(result.type).toEqual(MembersPage);
    expect(result.props.myProp).toEqual('property members');
  });
  it('Should return <DashboardPage /> with props when call renderLayoutDashboard', () => {
    const customProps = {
      myProp: 'property dashboard',
    };
    const result = renderLayoutDashboard({ ...customProps });
    expect(result.type).toEqual(DashboardPage);
    expect(result.props.myProp).toEqual('property dashboard');
  });
  it('Should return <FormPage /> with props when call renderLayoutForm', () => {
    const customProps = {
      myProp: 'property form',
    };
    const result = renderLayoutForm({ ...customProps });
    expect(result.type).toEqual(FormPage);
    expect(result.props.myProp).toEqual('property form');
  });
});
