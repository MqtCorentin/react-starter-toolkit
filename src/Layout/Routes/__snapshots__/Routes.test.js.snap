// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Route NotFound Page 1`] = `
<DocumentFragment>
  <div
    class="af-title-bar af-title-bar--hasstepper af-title-bar--backhome"
  >
    <div
      class="container af-title-bar__wrapper"
    >
      <h1
        class="af-title-bar__title"
      >
        title
      </h1>
      <a
        class="btn af-btn--circle"
        href="/"
        title="Retour à l'accueil"
      >
        <i
          class="glyphicon glyphicon-home"
        />
      </a>
      404 Page Not Found
    </div>
  </div>
  <div
    class="container af-container--notfound"
  >
    <h1
      class="af-notfound__title"
    >
      <div
        class="af-notfound__title-covernumber"
      >
        <span
          class="af-notfound__title-number"
        >
          404
        </span>
      </div>
      <div
        class="af-notfound__title-covernot"
      >
        <span
          class="af-notfound__title-not"
        >
          not
          <br />
          found
        </span>
      </div>
    </h1>
    <p
      class="af-notfound__message"
    >
      The page you are looking for is not here!
    </p>
  </div>
</DocumentFragment>
`;

exports[`Route Should render Routes 1`] = `
<DocumentFragment>
  <div
    style="min-height: 94px; position: relative;"
  >
     
  </div>
  <div
    class="af-nav-container"
  >
    <button
      aria-controls="mainmenu"
      aria-haspopup="true"
      class="mask fade "
      id="togglemenu"
      type="button"
    />
    <nav
      aria-label="Menu principal"
      class="af-nav af-drawer left "
      role="navigation"
    >
      <div
        class="container af-nav__wrapper"
      >
        <div
          class="af-nav__aside-header"
        >
          <h4
            class="af-nav__aside-title"
          >
            Menu
          </h4>
          <button
            aria-controls="mainmenu"
            aria-haspopup="true"
            class="af-nav__aside-close"
            id="togglemenu"
            type="button"
          >
            ×
          </button>
        </div>
        <ul
          aria-labelledby="togglemenu"
          class="af-nav__list "
          id="mainmenu"
          role="menubar"
        >
          <div
            style="min-height: 94px; position: relative;"
          >
             
          </div>
          <div
            style="min-height: 94px; position: relative;"
          >
             
          </div>
          <div
            style="min-height: 94px; position: relative;"
          >
             
          </div>
        </ul>
      </div>
    </nav>
  </div>
  <a
    aria-controls="mainmenu"
    aria-haspopup="true"
    class="btn af-title-bar__mobile-menu af-btn--circle"
    href="#"
    id="togglemenu"
    role="button"
    tabindex="0"
    title="Toggle menu"
  >
    <i
      class="glyphicon glyphicon-menu-hamburger"
    />
  </a>
  <div
    class="af-title-bar"
  >
    <div
      class="container af-title-bar__wrapper"
    >
      <h1
        class="af-title-bar__title"
      >
        Accueil
      </h1>
    </div>
  </div>
  <div
    class="af-main container"
  >
    <h1
      class="af-title--content"
    >
      Tableau de bord
    </h1>
  </div>
  <footer
    class="af-footer"
  >
    <div
      class="container-fluid container"
    >
      <a
        class="af-logo"
        href="https://www.axa.fr/"
        target="blank"
        title="Site Axa"
      >
        <img
          alt="Logo Axa"
          class="af-logo__brand"
          src="logo-axa.svg"
        />
      </a>
      <div
        class="af-footer-content"
      >
        © 2020 AXA Tous droits réservés - v0.1.0
      </div>
    </div>
  </footer>
</DocumentFragment>
`;
